### ISSUE TITLE
False positive: fetch is sometimes required for local resource-to-dataURL conversion

### ISSUE THREAD

**Issue by zsviczian** [obsidianmd/eslint-plugin#176]
The rule `Unexpected use of 'fetch'. Use the built-in requestUrl function instead of fetch for network requests in Obsidian.` is generally good guidance, but it is too broad in this case.

In the case linked below, `fetch` is not being used as a generic network request helper. It is used to load a local font file and convert it into a base64 data URI so it can be embedded into an SVG loaded as a `data:` URL. This is necessary because `@font-face` references via relative or `app://` URLs are blocked in that scenario, and `requestUrl` is not a suitable replacement for this use case.

https://github.com/zsviczian/obsidian-excalidraw-plugin/blob/9c034b9f51b63a7018785c9a170de26d8f22ff42/src/shared/EmbeddedFileLoader.ts#L1775-L1808

Since a linter cannot reliably infer intent from every `fetch` call, and this is one of the cases where the usage is valid despite the rule, the rule should either allow this pattern or provide an escape hatch for cases where `fetch` is intentionally used to obtain binary data for local asset embedding. Ideally, we should have a more context-aware rule so it only flags `fetch` when it is clearly being used as a generic remote request instead of asset loading or binary conversion.

More broadly, warnings need to be actionable or suppressible at the scanner level, not just locally in code. Otherwise, the signal-to-noise ratio drops, the recommendation becomes ambiguous, and over time the rule loses its relevance.

**Comment by saberzero1** [comment]
Question regarding the function you linked. You state that it is not using `fetch` as a generic network helper, but based on the function structure, it looks like it is possible to use it as such.

```ts
const fetchFontAsDataURI = async (url: string): Promise<string | null> => {
  if (url.startsWith("data:")) return url;
  try {
    const res = await fetch(url); // arbitrary network request
    if (!res.ok) return null;
    // ...rest of function...
```

I understand that the rest of the function the processes and handles the received data, but it still looks like a function that also accepts network paths. The linter cannot easily determine intent like this; It sees a `fetch` call and flags accordingly.

Perhaps [readBinary](https://docs.obsidian.md/Reference/TypeScript+API/DataAdapter/readBinary) (optionally in combination with [normalizePath](https://docs.obsidian.md/Reference/TypeScript+API/normalizePath)) is a better fit for local font fetching. [getResourcePath](https://docs.obsidian.md/Reference/TypeScript+API/Vault/getResourcePath) could help in the URI case.

Regarding the actionability of warnings. I agree that flagged issues should be actionable, but sometimes that action might be "don't use this." Obviously, we prefer to keep such cases to a minimum, and the rules that cannot be disabled are by and large related to security.

Can you please elaborate what makes [requestUrl](https://docs.obsidian.md/Reference/TypeScript+API/requestUrl) (or the suggestions above) insufficient for your use-case?

**Comment by hesprs** [comment]
@saberzero1 `fetch()` combined with `getResourcePath()` for a local file **does have advantages**. As shown in [this comment](https://forum.obsidian.md/t/add-api-support-for-reading-and-writing-binary-by-chunks-or-streaming/77384/12), `fetch()` allows streaming a local file, extremely essential for large file handling. Currently Obsidian doesn't provide any API for file read chunking or streaming.

**Comment by saberzero1** [comment]
> @saberzero1 `fetch()` combined with `getResourcePath()` for a local file **does have advantages**. As shown in [this comment](https://forum.obsidian.md/t/add-api-support-for-reading-and-writing-binary-by-chunks-or-streaming/77384/12), `fetch()` allows streaming a local file, extremely essential for large file handling. Currently Obsidian doesn't provide any API for file read chunking or streaming.

Perhaps I am misunderstanding, but the example provides a stream writer a describes that `fetch()` can be used for reading instead of a stream reader. This implies that while `fetch()` is an option, a stream reader implementation is also possible.

Wouldn't Obsidian's [readBinary](https://docs.obsidian.md/Reference/TypeScript+API/DataAdapter/readBinary) or the [Streams API](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API) provide this functionality?

As a side note, another user has flagged that the `requestUrl()` might have issues that make it behave differently that the definition implies. This has been reported to the Obsidian team directly. I currently have no status update on this.

**Comment by hesprs** [comment]
Obsidian's `readBinary` reads the entire file at once without streaming, when you read a 1GB file, the memory consumption spikes by 1GB, and the APP will crash very possibly.

The Streams API only processes a stream, they cannot read an Obsidian vault file in stream. Also, Node stream API only exist on desktop, so they are filtered off for mobile compatibility.

According to my current observation, only `fetch()` + `getResourcePath()` can reliably stream a local file (which sends the file in continuous small chunks, and garbage collected immediately when used) without exhausting the device's memory.

**Comment by saberzero1** [comment]
> Obsidian's `readBinary` reads the entire file at once without streaming, when you read a 1GB file, the memory consumption spikes by 1GB, and the APP will crash very possibly.
> 
> The Streams API only processes a stream, they cannot read an Obsidian vault file in stream. Also, Node stream API only exist on desktop, so they are filtered off for mobile compatibility.
> 
> According to my current observation, only `fetch()` + `getResourcePath()` can reliably stream a local file (which sends the file in continuous small chunks, and garbage collected immediately when used) without exhausting the device's memory.

In your linked comment you describe that you don't *need* `readBinary`. My question is whether it it possible to stream via `readBinary`, which your linked comment seems to imply that it is possible.

**Comment by hesprs** [comment]
It is impossible, my previous comment contains a proposed API about `readBinary` for chunking. But there is so such API.

**Comment by saberzero1** [comment]
> It is impossible, my previous comment contains a proposed API about `readBinary` for chunking. But there is so such API.

Alright, but that is a feature request for the Obsidian Forums, not a bug report for the eslint plugin. I understood that `readBinary` works for the case of loading a local file, but might OOM on larger files. `fetch` is not strictly required to load a file from disk, which is what this relevant for this issue.
